File: mergeSort.c
#include <stdio.h>
#include "mergeSort.h"
int *list;
int n;

void devide(int , int);
void merge(int, int, int);

void mergesort(int x[20], int y) {
    list = x;
    n = y;
    devide(0, n-1);
}

void devide(int p, int r) {
    int q;
    if(p<r) {
        q = (p+r)/2;
        devide(p, q);
        devide(q+1, r);
        merge(p, q, r);
    }
}

void merge(int p, int q, int r) {
    int L[20],R[20], i,j,k,x,y;

    x = 0;
    for(i=p; i<=q; i++){
        L[x] =  list[i];
        x++;
    }

    y = 0;
    for(j=q+1; j<=r; j++) {
        R[y] = list[j];
        y++;
    }

    i = j = 0;
    k = p;
    while(i<x && j<y) {
        if(L[i] > R[j]) {
            list[k] = R[j];
            j++;
        }else {
            list[k] = L[i];
            i++;
        }
        k++;
    }

    while(i<x) {
        list[k] = L[i];
        k++;
        i++;
    }

    while(j<y) {
        list[k] = R[j];
        k++;
        j++;
    }
}

File: mergeSort.java
public class mergeSort extends sortArray{
    public void sort(int[] list) {
        int p,r;
        p = 0;
        r = list.length;

        devide(list, p, r-1);
    }

    private static void devide(int[] x, int p, int r) {
        int q = (p + r) / 2;
        if(r > q) {
            devide(x, p, q);
            devide(x, q+1, r);
            merge(x, p, q, r);
        }
    }

    private static void merge(int[] x, int p, int q, int r) {
        int m, n;
        m = q-p+1;
        n = r-q;

        int[] L,R;
        L = new int[m];
        R = new int[n];

        for(int i=0; i<m; i++) {
            L[i] = x[p+i];
        }

        for(int i=0; i<n; i++) {
            R[i] = x[q+1+i];
        }

        int i=0, j=0, k=p;
        while(i<m && j<n) {
            if(L[i] <= R[j]){
                x[k] = L[i];
                i++;
            }else{
                x[k] = R[j];
                j++;
            }
            k++;
        }

        while(i<m) {
            x[k] = L[i];
            i++;
            k++;
        }

        while(j<n) {
            x[k] = R[j];
            j++;
            k++;
        }
    }
}
